tests/test_unicode.py

```diff
@@ -0,0 +1,62 @@
+import unittest
+from unicodedata import normalize
+
+from hypothesis import given, strategies as st
+
+# For every (n1, n2, n3) triple, applying n1 then n2 must be the same
+# as applying n3.
+# Reference: http://unicode.org/reports/tr15/#Design_Goals
+compositions = [
+    ("NFC", "NFC", "NFC"),
+    ("NFC", "NFD", "NFD"),
+    ("NFC", "NFKC", "NFKC"),
+    ("NFC", "NFKD", "NFKD"),
+    ("NFD", "NFC", "NFC"),
+    ("NFD", "NFD", "NFD"),
+    ("NFD", "NFKC", "NFKC"),
+    ("NFD", "NFKD", "NFKD"),
+    ("NFKC", "NFC", "NFKC"),
+    ("NFKC", "NFD", "NFKD"),
+    ("NFKC", "NFKC", "NFKC"),
+    ("NFKC", "NFKD", "NFKD"),
+    ("NFKD", "NFC", "NFKC"),
+    ("NFKD", "NFD", "NFKD"),
+    ("NFKD", "NFKC", "NFKC"),
+    ("NFKD", "NFKD", "NFKD"),
+]
+
+
+class TestUnicode(unittest.TestCase):
+    @given(s=st.text(), comps=st.sampled_from(compositions))
+    def test_composition(self, s, comps):
+        # see issues https://foss.heptapod.net/pypy/pypy/issues/2289
+        # and https://bugs.python.org/issue26917
+        norm1, norm2, norm3 = comps
+        self.assertEqual(normalize(norm2, normalize(norm1, s)), normalize(norm3, s))
+
+    @given(u=st.text(), prefix=st.text(), suffix=st.text())
+    def test_find_index(self, u, prefix, suffix):
+        s = prefix + u + suffix
+        index = s.find(u)
+        index2 = s.index(u)
+        self.assertEqual(index, index2)
+        self.assertLessEqual(0, index)
+        self.assertLessEqual(index, len(prefix))
```

Comment:
I think this is only certainly true if we add `assume(u not in prefix)`, and a similar clause for suffix below.

Reply:
why? we check that `0 <= index <= len(prefix)`. if `u not in prefix`, then `index == len(prefix)`. Otherwise, `index` will be some other value smaller than `len(prefix)`, but definitely non-negative. I am going to try to add a comment that explains it, other readers will have the same question.

(other argument: if this were wrong, hypothesis would have found counterexamples by now ;-) )

Reply:
No, my reasoning is wrong (but `0 <= index <= len(prefix)` is still true) :grinning: 

Even if `u not in prefix`, it can happen that index < len(prefix), eg for `prefix = "0", u = "00"`

Reply:
(hypothesis found this counterexample, of course)

Reply:
I've now pushed a comment, does that help?

Reply:
Ah, yep... I think I was confusing the `index` and `rindex` cases, but the comment definitely helps.  Thanks!